File: ai/src/main/java/org/pingel/gestalt/core/SimpleTransformCall.java
package org.pingel.gestalt.core;

import java.util.Iterator;
import java.util.Map;
import java.util.TreeMap;

public class SimpleTransformCall extends CallGraph
{
	private SimpleTransform transform;

    public SimpleTransformCall(int id, History history, Lexicon lexicon, SimpleTransform transform, TransformEdge macro)
    {
    	super(id, history, lexicon, transform, macro);

    	this.transform = transform;
    	// TODO assert: out.variables equals bindings.keySet
    }

    public Map<Name, Form> replacements()
    {
		GLogger.global.entering("SimpleTransformCall", "replacements");

		Map<Name, Form> replacements = new TreeMap<Name, Form>();
		Iterator<Name> map_it = transform.map.keySet().iterator();
		while( map_it.hasNext() ) {

		    Name mi = map_it.next();
		    Name variable = transform.map.get(mi);
		    Form replacement = unifier.get(mi);

		    {
				// too strict !!
				Form already = replacements.get(variable);
				if( already != null ) {
				    GLogger.global.info("SimpleSystmCall.replacements variable is already bound to " + already.toString());
			    
					// TODO Shouldn't this have already been done in the Call superclass ??
	
					// matched = false;
					hasNext = false;
					return null;
				}
		    }

		    replacements.put(variable, replacement);
		}

        return replacements;
    }

    public void next(History history, Lexicon lexicon)
    {
		GLogger.global.entering("SimpleSystmCall", "next");

        Map<Name, Form> replacements = replacements();
        if( replacements == null ) {
            return;
        }

		Form resultForm = lexicon.getForm(transform.outName).duplicateAndReplace(replacements);
	
		CallVertex resultVertex = getGraph().addVertex(new CallVertex(history.nextVertexId(), transform.exitNode, resultForm));

		GLogger.global.info("Transform result is " + resultForm.toString());

		getGraph().addEdge(new CallEdge(history.nextEdgeId(), start, resultVertex, macroEdge));

		outputs.add(resultVertex);
		
		hasNext = false;

        GLogger.global.exiting("SimpleTransformCall", "next");
    }

    public double cost()
    {
        return transform.cost;
    }

}

File: gestalt/src/main/scala/org/pingel/gestalt/core/SimpleTransformCall.scala
package org.pingel.gestalt.core

case class SimpleTransformCall(
    override val id: Int,
    override val history: History,
    override val lexicon: Lexicon,
    override val transform: SimpleTransform,
    macro: TransformEdge)
extends CallGraph(id, history, lexicon, transform, macro)
{
	// TODO assert: out.variables equals bindings.keySet

    def replacements(): Map[Name, Form] = {
		GLogger.global.entering("SimpleTransformCall", "replacements")
		var replacements = Map[Name, Form]()
		var map_it = transform.map.keySet.iterator
		while( map_it.hasNext ) {
		    val mi = map_it.next()
		    val variable = transform.map(mi)
		    val replacement = unifier.get(mi)

		    // too strict !!
		    val already = replacements(variable)
		    if( already != null ) {
		    	GLogger.global.info("SimpleSystmCall.replacements variable is already bound to " + already.toString())
		    	// TODO Shouldn't this have already been done in the Call superclass ??
		    	// matched = false;
		    	_hasNext = false
		    	return null
		    }

		    replacements += variable -> replacement
		}

        replacements
    }

    def next(history: History, lexicon: Lexicon): Unit = {
		GLogger.global.entering("SimpleSystmCall", "next")
        val repls = replacements()
        if( repls == null ) {
            return
        }
		val resultForm = lexicon.getForm(transform.outName).duplicateAndReplace(repls)
		val resultVertex = addVertex(new CallVertex(history.nextVertexId(), transform.exitNode, resultForm))
		GLogger.global.info("Transform result is " + resultForm.toString())
		addEdge(new CallEdge(history.nextEdgeId(), start, resultVertex, macroEdge))
		outputs += resultVertex
		_hasNext = false
        GLogger.global.exiting("SimpleTransformCall", "next")
    }

    def cost() = transform.cost

}
